optaplanner-core/src/test/java/org/optaplanner/core/config/solver/EnvironmentModeTest.java

```diff
@@ -0,0 +1,237 @@
+        }
+    }
+
+    @Test
+    public void corruptedCustomMovesTest() {
+        // intrusive modes should throw exception about corrupted undoMove
+        setSolverConfigCalculatorClass(TestdataDifferentValuesCalculator.class);
+
+        switch (environmentMode) {
+            case FULL_ASSERT:
+            case FAST_ASSERT:
+                setSolverConfigMoveListFactoryClassToCorrupted(TestdataChangeMoveWithCorruptedUndoMoveFactory.class);
+                assertIllegalStateExceptionWhileSolving("corrupted undoMove");
+                break;
+            case NON_INTRUSIVE_FULL_ASSERT:
+                setSolverConfigMoveListFactoryClassToCorrupted(TestdataCorruptedEntityUndoMoveFactory.class);
+                assertIllegalStateExceptionWhileSolving("not the uncorruptedScore");
+                break;
+            case REPRODUCIBLE:
+            case NON_REPRODUCIBLE:
+            case PRODUCTION:
+                // no exception expected
+                break;
+            default:
+                throw new AssertionError("Missing case for " + environmentMode);
+        }
+    }
+
+    @Test
+    public void corruptedScoreRulesTest() {
+        // for full assert modes it should throw exception about corrupted score
+        setSolverConfigCalculatorClass(TestdataCorruptedDifferentValuesCalculator.class);
+
+        switch (environmentMode) {
+            case FULL_ASSERT:
+            case NON_INTRUSIVE_FULL_ASSERT:
+                assertIllegalStateExceptionWhileSolving("not the uncorruptedScore");
+                break;
+            case FAST_ASSERT:
+                assertIllegalStateExceptionWhileSolving("Score corruption analysis could not be generated ");
+                break;
+            case REPRODUCIBLE:
+            case NON_REPRODUCIBLE:
+            case PRODUCTION:
+                // no exception expected
+                break;
+            default:
+                throw new AssertionError("Missing case for " + environmentMode);
+        }
+    }
+
+    private void assertReproducibility(Solver solver1, Solver solver2) {
+        assertThat(areGeneratingSameNumbers(
+                ((DefaultSolver) solver1).getRandomFactory(),
+                ((DefaultSolver) solver2).getRandomFactory())
+        )
+                .as("Random factories generate different random values. This is 100% wrong.")
+                .isTrue();
+        assertThat(areScoreSeriesTheSame(solver1, solver2))
+                .as("Score steps are different and should be the same!")
+                .isTrue();
+    }
+
+    private void assertNonReproducibility(Solver solver1, Solver solver2) {
+        assertThat(areGeneratingSameNumbers(
+                ((DefaultSolver) solver1).getRandomFactory(),
+                ((DefaultSolver) solver2).getRandomFactory())
+        )
+                .as("Random factories generate exactly same results. "
+                            + "It can happen but the probability is very low. Run test again")
```

Comment:
This test is intentionally flaky? Not sure we should include it.

Reply:
I made the problem more complex and, since it being rather fast, had it running in the background while working on this PR today for a while. There were no failed tests out of about 120,000 runs. The probability of it failing is very low and would suggest keeping it as it is. What do you think?

Reply:
OK. But if I see it failing a build, I'll get back to you. :-)

Reply:
Deal! :)